PREFORMA: MediaConch integration

jrwdunham opened the pull request:
- Pre-ingest conformance: can check whether .mkv files are in
  conformance with the .mkv specification (implementation check).

- Post-normalization conformance: can check whether files normalized
  to .mkv are in conformance with the .mkv specification.

- Transfer policy check: check whether original files conform to
  a pre-defined MediaConch policy.

- Ingest policy check: can check that derivatives
  (normalized for access or preservation, manually or AM-created)
  conform to a pre-defined policy.

sevein commented on src/MCPServer/share/sharedDirectoryStructure/sharedMicroServiceTasksConfigs/policies/.keep:
```diff
@@ -0,0 +1,2 @@
+This file exists just to ensure that git will add this (otherwise empty)
```
Our convention to track empty directories in git is to leave a `.gitignore` file like the following: https://goo.gl/RN90mS. I don't know if it's the best approach toady but perhaps we should be consistent. More here: http://stackoverflow.com/a/11579945.

jrwdunham replied:
Hey, who you callin' toady! Just kidding: change effected.

sevein commented on src/MCPClient/lib/clientScripts/policyCheck.py:
```diff
@@ -0,0 +1,331 @@
+FAIL_CODE = 1
+
+
+class PolicyChecker:
```
Should your new class subclass `object`? I think we always use the new style.

jrwdunham replied:
Yes, you're right. I'm used to writing Python3-only code where subclassing `object` is redundant.

sevein commented on src/MCPClient/lib/clientScripts/validateDerivative.py:
```diff
@@ -0,0 +1,179 @@
+NO_RULES_CODE = 0
+
+
+class DerivativeValidator:
```
Same here, I think we want to subclass `object`.

jrwdunham replied:
Ok.

sevein commented on src/MCPClient/lib/clientScripts/validateDerivative.py:
```diff
@@ -0,0 +1,179 @@
+from __future__ import print_function
```
It seems that `validateDerivative.py` is not an executable script but your commit includes the file with the permission bit on it. This is not the only case I think, e.g. `transcoder.py` has both a shebang and the execution bit - we should take care of those some other time, I'll file an issue.

I think I would prefer to put files that are not entry points somewhere else. The easiest way could be a `lib/` folder inside `clientScripts`. This is not strictly required but I think it would be nice. I like your class-based approach. Some of these classes like `PolicyChecker` could as well go to `lib`?

sevein replied:
@jhsimpson @Hwesta, what do you think?

jrwdunham replied:
Yes, assuming the clientScripts/lib/ approach is what we go with, the entire policyCheck.py could be moved there too and made non-executable. It's not really used as a client script on its own; it's always imported and subclassed just like validateDerivative.py

sevein commented on src/dashboard/src/components/administration/urls.py:
```diff
@@ -39,6 +39,9 @@
     url(r'premis/agent/$', views.premis_agent),
     url(r'api/$', views.api),
+    url(r'policies/$', views.policies),
+    url(r'policies/view/(?P<policy>[^/]+)/$', views.view_policy),
```
Can we replace this one with `r'policies/(?P<policy>[^/]+)/$'`? It seems consistent with other URLs across the dashboard.

jrwdunham replied:
Ok, that makes sense.

sevein commented on src/dashboard/src/components/administration/urls.py:
```diff
@@ -39,6 +39,9 @@
     url(r'api/$', views.api),
+    url(r'policies/$', views.policies),
+    url(r'policies/view/(?P<policy>[^/]+)/$', views.view_policy),
+    url(r'policies/delete/(?P<policy>[^/]+)/$', views.delete_policy),
```
I realize we're not consistently naming our URLs, sometimes we use the `GET /foobar/action/$id/` format but in other cases we've been using `GET /foobar/$id/action/`. I guess we can discuss that some other day.

jrwdunham replied:
I guess I was just being consistent with what's done in this particular file.

sevein commented on src/dashboard/src/templates/administration/policies.html:
```diff
@@ -0,0 +1,79 @@
+          <div class='form-group'>
+            <label class='btn btn-default' for='file'>
+              <input id='file' type='file' name='file' style='display:none;'
+                onchange='$("#filename").html($(this).val().replace("C:\\fakepath\\", ""));'>
```
What is the routine in `onchange` for?

jrwdunham replied:
When the user selects a file to upload, the `<span#filename>` displays the filename. Some browsers list this "fakepath" stuff and it looks ugly, so the `replace` just removes it. See http://stackoverflow.com/questions/6365858/use-jquery-to-get-the-file-inputs-selected-filename-without-the-path

sevein commented on src/dashboard/src/templates/administration/policy.html:
```diff
@@ -0,0 +1,60 @@
+      white-space: pre;
+      line-height: 2em;
+      background-color: transparent;
+    }
```
This `<style>...</style>` is the same you have above in the same file excepting this extra `background-color` here.

jrwdunham replied:
Yeah, that's odd I did that. I'll merge into a single <style> block.

sevein commented on src/dashboard/src/templates/administration/policy.html:
```diff
@@ -0,0 +1,60 @@
+  </div>
+
+{% endblock %}
+
```
This blank line is not needed.

sevein commented on src/dashboard/src/components/administration/forms.py:
```diff
@@ -258,6 +265,20 @@ class ProcessingConfigurationForm(forms.Form):
+    processing_fields['8ce07e94-6130-4987-96f0-2399ad45c5c2'] = {
+        'type': 'boolean',
+        'name': 'policy_checks_access_derivatives',
+        'label': 'Perform policy checks on access derivatives',
```
Justin mentioned that this should probably be renamed to `Validate access derivatives`, similar with the other cases. We can talk about this later.

jrwdunham replied:
This proposed change seems confusing to me since this decision point governs whether "policyCheckAccessDerivative.py" gets called and there is a distinct client script named "validateAccessDerivative.py" that has nothing to do with this...

sevein reviewed (commented):
@jrwdunham I've got a few observations for you, see my comments inline.

Some of the stuff you're doing in this pull request is interacting with parts of the application that I don't really know very well. I noticed that some of the changes you do in `0030_mediaconch.py` belong in `fpr-admin` and the scripts under `migrations-data` should perhaps be moved to https://github.com/artefactual/archivematica-fpr-tools. Justin also mentioned a couple of things that need to be discussed, like the labels used in the processing configuration page may have to be reworded and some stuff about the validation tools I can't remember well! He's going to either participate in this pull request or schedule a meeting to discuss this IRL. I'm interested to participate so I can get a better understanding on how the FPR works!

jrwdunham commented:
@sevein I take your point about the 0030_mediaconch.py migration potentially being better placed in the archivematica-fpr-admin repo. However, I'd like some clarification on this and a better high-level understanding of the motivations for this (from you or @Hwesta / @jhsimpson) before I go ahead and make changes. 

Some of the models I'm adding are AM models with classes defined in the dashboard and others are FPR models. I could move the entire migration to archivematica-fpr-admin, but some quick greping suggests that the migrations there never touch the AM models proper.

So would the idea be to split this migration across 2 different files in 2 different repos? That would seem to needlessly complicate things, especially given that this migration is conceptually atomic.

Relatedly, I could put the two new commands in archivematica-fpr-tools, but the "deployment process" in the README talks about generating raw SQL migrations, which doesn't seem to jive with how we currently do migrations (Django).

sevein commented on src/MCPClient/lib/clientScripts/validateFile.py:
```diff
@@ -1,91 +1,297 @@
+        return self._sip_pres_val_dir
+
+
+def _get_logger_name(file_type):
```
Is this still relevant?

jrwdunham replied:
No, it's not. Thanks. Will remove.

sevein commented on src/MCPClient/lib/clientScripts/validateFile.py:
```diff
@@ -1,91 +1,297 @@
-    sys.exit(main(file_path, file_uuid, sip_uuid))
+    shared_path = _get_shared_path(sys.argv)
+    file_type = _get_file_type(sys.argv)
+    logger = get_script_logger(_get_logger_name(file_type))
```
I think you should use the logger instead of `sys.stderr`. The events are currently [written to disk](https://github.com/artefactual/archivematica/blob/stable/1.6.x/src/archivematicaCommon/lib/custom_handlers.py#L20) but we're planning to change that soon.

sevein replied:
See #606.

sevein commented on src/MCPClient/requirements/base.txt:
```diff
@@ -1,3 +1,4 @@
+git+git://github.com/artefactual-labs/ammcpc@v0.1.0#egg=ammcpc
```
I tagged and updated this.

sevein commented on src/dashboard/src/requirements/base.txt:
```diff
@@ -9,7 +9,7 @@ django-extensions==1.1.1
 django-annoying==0.7.7
 elasticsearch>=1.0.0,<2.0.0
-git+https://github.com/artefactual/archivematica-fpr-admin.git@v1.3.0-beta.1#egg=archivematica-fpr-admin
+git+https://github.com/artefactual/archivematica-fpr-admin.git@dev/issue-9478-preforma#egg=archivematica-fpr-admin
```
Now in qa/1.x we import archivematica-fpr-admin using pip. I've updated this using your branch.

sevein commented on src/dashboard/src/main/migrations/0034_mediaconch_policy_checks.py:
```diff
@@ -0,0 +1,451 @@
+    validate_formats_cl.defaultnextchainlink = rgnl_policy_check_choice_cl
+    MicroServiceChainLinkExitCode.objects\
+        .filter(microservicechainlink=validate_formats_cl)\
+        .update(nextmicroservicechainlink=rgnl_policy_check_choice_cl)
```
With this you're leaving behind the old task config, right? Should we delete it?

sevein reviewed (approved):
@jrwdunham - please take a look at my changes. There is a minor observation about a chain link config left behind. Other than that this is 👍. Good job, I've learned a lot! The commits prefixed with #601 can be deleted, I needed them to run the test suite.

jrwdunham commented:
This PR should be closed since it is replace by https://github.com/artefactual/archivematica/pull/634. Before closing, I would like to make sure that the issues reference above are addressed or are being addressed:

- The commits prefixed with #601 that sevein mentions: are these still in existence? still relevant?
- What is the "observation about a chain link config left behind" that sevein mentions?
- Requesting more details on the suggestion to use the logger instead of sys.stderr, and whether that will affect what is displayed in the /tasks/<UUID> pages

sevein commented:
Can we delete these other two branches?
- dev/issue-9478-preforma-test
- dev/issue-9478-preforma-test-qa-1-x

jrwdunham committed:
MediaConch integration: validation & policy checks

- Pre-ingest validation can check whether .mkv files are in conformance with
  the .mkv specification (implementation check).
- Post-normalization validation: can check whether files normalized
  to .mkv are in conformance with the .mkv specification.
- Transfer policy check: check whether original files conform to a pre-defined
  MediaConch policy.
- Ingest policy check: can check whether derivatives (normalized for access or
  preservation, manually or AM-created) conform to a pre-defined policy.
- Uses ammcpc (https://github.com/artefactual-labs/ammcpc) in policy check
  command
